packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/watch-blocks/watch-blocks-v1-endpoint.ts

```diff
@@ -0,0 +1,329 @@
+      return;
+    }
+
+    socket.emit(WatchBlocksV1.Next, {
+      fullBlock: blockEvent,
+    });
+  }
+
+  /**
+   * Callback executed when receiving block with standard type "filtered"
+   * Sends WatchBlocksV1.Next with new block to the client.
+   *
+   * @param blockEvent filtered block
+   *
+   * @returns Nothing.
+   */
+  private monitorFilteredCallback(blockEvent: BlockEvent) {
+    const { socket, log } = this;
+    const clientId = socket.id;
+    log.debug(
+      `Filtered BlockEvent received: #${blockEvent.blockNumber.toString()}, client: ${clientId}`,
+    );
+
+    if (!("filtered_transactions" in blockEvent.blockData)) {
+      log.error("Wrong blockEvent type received - should not happen!");
+      return;
+    }
+
+    socket.emit(WatchBlocksV1.Next, {
+      filteredBlock: blockEvent,
+    });
+  }
+
+  /**
+   * Callback executed when receiving block with standard type "private"
+   * Sends WatchBlocksV1.Next with new block to the client.
+   *
+   * @param blockEvent private block
+   *
+   * @returns Nothing.
+   */
+  private monitorPrivateCallback(blockEvent: BlockEvent) {
+    const { socket, log } = this;
+    const clientId = socket.id;
+    log.debug(
+      `Private BlockEvent received: #${blockEvent.blockNumber.toString()}, client: ${clientId}`,
+    );
+
+    if (!("data" in blockEvent.blockData)) {
+      log.error("Wrong blockEvent type received - should not happen!");
+      return;
+    }
+
+    socket.emit(WatchBlocksV1.Next, {
+      privateBlock: blockEvent,
+    });
+  }
+
+  /**
+   * Get block listener callback and listener type it's expect.
+   * Returns separate function object each time it's called (this is required y fabric node SDK).
+   *
+   * @param type requested listener type (including custom Cactus ones).
+   *
+   * @returns listener: BlockListener;
+   * @returns listenerType: BlockType;
+   */
+  private getBlockListener(type: WatchBlocksListenerTypeV1) {
+    let listener: BlockListener;
+    let listenerType: EventType;
+
+    switch (type) {
+      case WatchBlocksListenerTypeV1.Full:
+        listener = async (blockEvent) => this.monitorFullCallback(blockEvent);
+        listenerType = "full";
+        break;
+      case WatchBlocksListenerTypeV1.Filtered:
+        listener = async (blockEvent) =>
+          this.monitorFilteredCallback(blockEvent);
+        listenerType = "filtered";
+        break;
+      case WatchBlocksListenerTypeV1.Private:
+        listener = async (blockEvent) =>
+          this.monitorPrivateCallback(blockEvent);
+        listenerType = "private";
+        break;
+      case WatchBlocksListenerTypeV1.CactusTransactions:
+        listener = async (blockEvent) =>
+          this.monitorCactusTransactionsCallback(blockEvent);
+        listenerType = "full";
+        break;
+      default:
+        // Will not compile if any type was not handled by above switch.
+        const unknownType: never = type;
```

Comment:
@outSH Why do you need to cast it to `never`? Couldn't you just log the value of `type` instead of this new `unknownType` variable?

Reply:
@petermetz There's a comment above - Will not compile if any type was not handled by above switch. It's not casting, type in default is already of `never` type if the switch check above is exhaustive. If, in the future, someone adds another type `Foo`, then type in default will be `Foo` (since it wasn't narrowed by any case above), and assigning this to never will break compilation - marking place where type handling has to be added. The log below is to catch type unexpected by compiler, and also make use of `unknownType` which would cause warning otherwise :)

So this is just a trick for additional compile time check.

Reply:
@outSH 

> There's a comment above - Will not compile if any type was not handled by above switch.

Yeah that's fine, I'm not saying we should remove the `default` block, I'm saying that this weird casting of the `type` variable doesn't do anything useful (as far as I can tell but could be totally wrong 'coz I do that a lot anyway)

> If, in the future, someone adds another type Foo, then type in default will be Foo (since it wasn't narrowed by any case above), and assigning this to never will break compilation - marking place where type handling has to be added.

> So this is just a trick for additional compile time check.

I agree with all of that, 100%.

What I'm saying is that something like this also works as the code for the default branch and it is less confusing because there is no weird casting (and I added this extra little thing that it dumps the valid enum values back into the error message so that people can know what their options were)

```typescript
        const validTypes = Object.keys(WatchBlocksListenerTypeV1).join(";");
        const errorMessage = `Unknown block listen type '${type}'. Check name and connector version. Accepted listener types for WatchBlocksListenerTypeV1 are: [${validTypes}]`;
        throw new RuntimeError(errorMessage);
```

which then produces error messages like this:

```sh
[2022-12-19T20:15:50.069Z] ERROR (WatchBlocksV1Endpoint): RuntimeError: Unknown block listen type 'Some_INVALID_WatchBlocksListenerTypeV1'. Check name and connector version. Accepted listener types for WatchBlocksListenerTypeV1 are: [Filtered;Full;Private;CactusTransactions]
        at WatchBlocksV1Endpoint.getBlockListener (/***/cactus/packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/watch-blocks/watch-blocks-v1-endpoint.ts:260:15)
```

I tried to write a test for it as well (please see my suggested patch that you can apply with one click from VSCode via the GitHub extension) but the test case itself does not work because if I feed an invalid enum value to the watch endpoint the `testWatchBlock()` test utility method does not seem to complete the promise at all (e.g. neither reject nor resolve).

Please take a look at the patch and let me know what you think. If I'm just not understanding something then it's fine to merge as-is too because I don't want to put extra work/burden on you.

This is the patch comment I'm talking about: https://github.com/hyperledger/cactus/pull/2119#issuecomment-1358266362

Reply:
@petermetz 
Your proposal is extending the runtime error handling (i.e. we received some unknown value and want to handle it accordingly). My solution (with assignment to `never`) is to handle issue of not including specific option in switch during compilation, and throw an error during runtime.

[Have a look at this TS playground example I've prepared](https://www.typescriptlang.org/play?#code/KYOwrgtgBAyg9hYBRc0DeBYAUFXUCCUAvFAOQCGpANNnlAELFkBG1teAwk6QMZtYBfbNgBmYEDwAuASzggowAB4ALcmADOMgG7AOy4DwDWACkkBPAA7AAXLATJUASiiYcUdQHdpknstOXgZ1c6HnJ1YDtEFEgAOnxrdhC5dTgAG2AY1LgAc2MAIgBxOEkCPMdEvGYAJ2ByQwBuCtDwyIdY+gS3JJAU9Myc-KKS+jKK3GrahqawiPgo1BiOTrpcHmS0jKzcwuKoDlGuypq6xsOAE2ARNVTJZZXV5JLxQxA4DxAAFQDbEGAdKqY5isp3uuEkyiqbygvw8UCQVUhVWMYxWAAMAKogF5veRAiIAWjIABI0M9Xu8vlYBKRUTRDnRHCCoEJBMIsGIJDI5NDivAkCo1JppDo9AYTHjbHM2hAguxPN5fP4rLLDs1ZvZohA4ndur1NgMdiV8AcQjNWpqYh0UWseht+tshgwTXgLlcwDcdSsbZooGScZSbNC-sAASQ8fUoAB6SNQANkKWa0hQaTqHklMLqaTZEDkZjpKCSOAFgJkX7-UgxYwAJgAzFWq+V6XhwZDYTC4Qi4EiUXQMVjybiS4TSCS-RSAtTaT2oIz2CyWaJxFJZPIICnwmd6Mdxd9zU4XHKvD4-HiVaaWgmFvFres+ltBrtjY37hMTtMLxqFlam967ffDU6z4rK+Uyqmal6xEsN62neBqOvsQF0NGIFMrgrrXLc07ek8-b+ruZYhoCASoSsLZQu28KIsiTa9pi2LvMWVhQMOo64eOVI0nSoIzky85AA)

- `exhaustiveCheck` is working code, all cases are handled.
- `notSoExhaustiveCheck` is missing handling of one type (for example - we've added a new type and forgot about switch here). Assignment to `unknownType` fails because rhs is not `never` - there are cases it will have valid value we just don't support yet.
- `missedBreak` example show another kind of error - we didn't break in single case and it's possible to fall to default by accident. This is also catched.

Value not defined in our definitions is handled correctly in all cases.

Reply:
@petermetz I've included `RuntimeError` and test you've proposed, but kept assignment to `never` because of reasons described above.

Reply:
@outSH Whoah, epic facepalm. I finally get it, thank you for the patience and the detailed explanation! 
Damn, I really thought the TS compiler was preventing unhandled enum values by default without this (seemingly) pointless extra `never` cast thingy. Sorry and thanks again!